src/test/java/com/google/sps/servlets/UpdateServletTest.java

```diff
@@ -164,4 +173,60 @@ public void testSeveralDevices() throws IOException {
     verify(mockedUtil, times(1)).updateDevices(TEST_USER_ID, Arrays.asList("device1", "device2"), "{\"annotatedAssetId\":\"ABC123\"}");
   }
 
+  @Test
+  public void testOneDeviceFailed() throws IOException {
+    when(mockedUserService.isUserLoggedIn()).thenReturn(true);
+    when(mockedUserService.getCurrentUser()).thenReturn(userFake);
+
+    when(request.getParameter("annotatedLocation")).thenReturn(null);
+    when(request.getParameter("annotatedAssetId")).thenReturn("ABC123");
+    when(request.getParameter("annotatedUser")).thenReturn(null);
+    when(request.getParameter(servlet.DEVICE_IDS_PARAMETER_NAME)).thenReturn("[device1, device2]");
+
+    when(mockedUtil.updateDevices(TEST_USER_ID, Arrays.asList("device1", "device2"), "{\"annotatedAssetId\":\"ABC123\"}")).thenReturn(Arrays.asList("device1"));
+
+    servlet.doPost(request, response);
+
+    verify(response).sendRedirect(servlet.INDEX_URL);
+    verify(mockedUserService, times(1)).isUserLoggedIn();
+    verify(request, times(1)).getParameter("annotatedLocation");
+    verify(request, times(1)).getParameter("annotatedAssetId");
+    verify(request, times(1)).getParameter("annotatedUser");
+    verify(request, times(1)).getParameter(servlet.DEVICE_IDS_PARAMETER_NAME);
+    verify(mockedUtil, times(1)).updateDevices(TEST_USER_ID, Arrays.asList("device1", "device2"), "{\"annotatedAssetId\":\"ABC123\"}");
+    
+    verify(response).setContentType("application/json");
+    String result = stringWriter.getBuffer().toString().trim();
+    String expected = "[\"device1\"]";
+    Assert.assertEquals(expected, result);
+  }
+
+  @Test
+  public void testOnAllDeviceFailed() throws IOException {
+    when(mockedUserService.isUserLoggedIn()).thenReturn(true);
+    when(mockedUserService.getCurrentUser()).thenReturn(userFake);
+
+    when(request.getParameter("annotatedLocation")).thenReturn(null);
+    when(request.getParameter("annotatedAssetId")).thenReturn("ABC123");
+    when(request.getParameter("annotatedUser")).thenReturn(null);
+    when(request.getParameter(servlet.DEVICE_IDS_PARAMETER_NAME)).thenReturn("[device1, device2]");
+
+    when(mockedUtil.updateDevices(TEST_USER_ID, Arrays.asList("device1", "device2"), "{\"annotatedAssetId\":\"ABC123\"}")).thenReturn(Arrays.asList("device1","device2"));
+
+    servlet.doPost(request, response);
+
+    verify(response).sendRedirect(servlet.INDEX_URL);
+    verify(mockedUserService, times(1)).isUserLoggedIn();
+    verify(request, times(1)).getParameter("annotatedLocation");
+    verify(request, times(1)).getParameter("annotatedAssetId");
+    verify(request, times(1)).getParameter("annotatedUser");
+    verify(request, times(1)).getParameter(servlet.DEVICE_IDS_PARAMETER_NAME);
```

Comment:
These seem repeating a lot. Maybe we can put them in a helper function verifyUpdateExecutedAsExpected - you can come up with a better name :)

Reply:
will do